### PR TITLE
BYOI setAuthToken

### PR DESCRIPTION
**Pull request by guperrot**
Please have a look at our [guidelines for contributions](https://github.com/microsoft/appcenter-sdk-android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Add AppCenter.setAuthToken

## Related PRs or issues

[AB#67952](https://dev.azure.com/msmobilecenter/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/67952)

### CODE REVIEW

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterAuthTest.java [hunk 1]
```diff
@@ -42,6 +42,44 @@ public void setUp() throws Exception {
...
+    }
+
+    @Test
+    public void unsetAuthToken() {
+
+        /* When we unset auth token. */
+        AppCenter.setAuthToken(null);
+
+        /* Then it's removed. */
+        verify(mAuthTokenContext).setAuthToken(null, null, null);
```
Comment (MatkovIvan) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterAuthTest.java]: And `JwtClaims.parse` isn't called.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 2]
```diff
@@ -1170,6 +1182,24 @@ public void run() {
...
+    /**
+     * Implements {@link #setAuthToken} at instance level.
+     */
+    private void setInstanceAuthToken(String authToken) {
+        AuthTokenContext authTokenContext = AuthTokenContext.getInstance();
+        JwtClaims claims = null;
+        if (authToken != null) {
+            claims = JwtClaims.parse(authToken);
+        }
+        if (claims != null) {
```
Comment (jaelim-ms) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: You could invert the condition.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterAuthTest.java (hunk 1 above)
Comment (jaelim-ms) [reply sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterAuthTest.java]: I think it is fine but we can also verify that isn't called.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java (hunk 2 above)
Comment (jwallra) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: I prefer this organization, but I suppose it's a stylistic choice

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 3]
```diff
@@ -1170,6 +1182,24 @@ public void run() {
...
+            AppCenterLog.debug(LOG_TAG, "Token has been refreshed.");
```
Comment (jwallra) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: To match the other log in the method

```suggestion
            AppCenterLog.debug(LOG_TAG, "Authentication token has been refreshed.");
```

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java (hunk 2 above)
Comment (jaelim-ms) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: We can get rid of unnecessary one bit operation there. 😄

Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: Not really the logging matters here.
If you pass null we just log that you remove the token, it's not an error.
If you pass an invalid JWT, we log an error and utils AND we remove the token.

Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: It's more common to have the success block before the failure one in a if/else. Both work.

**Pull request merged**